### PR TITLE
Reservation user swapping

### PR DESCRIPTION
**Pull request by abuckles-uci**
Hello,

This is a customization we've patched in for use in our labs at UC Irvine. It allows for users to give their reservation to a different user (given certain conditions, such as what group the user is in and how much time there is before the reservation starts). Staff are allowed to disregard the conditions. This was useful for our users so they didn't have to delete their reservations if giving it to another user, which would risk losing the timeslot.

There are some improvements that could be made, such as removing some hard coded values and allowing for this feature to be toggled off/on.

image

### CODE REVIEW

**Comment by rptmat57** [comment]
Hello,
Thank you for submitting this, the idea of letting users transfer their reservation to other users is an interesting one!
I see this feature as pretty similar to reserving a tool on behalf of someone else (which currently only staff can do)

I have 2 main issues with the PR in its current state:

1. As you said, there are a lot of hardcoded group names, times, rules etc. and I understand that those make sense for your facility but they probably would not make sense in any other.
2. There are a certain number of ways to break NEMO rules by using this feature, which you probably are ok with for your facility but this is unlikely to be the case in other places:

    1. None of the policy rules are checked and that could cause issues. Also the receiving user does not get a say in this transaction.
    2. If the user doesn't know he is receiving the reservation, and misses it, he could potentially be charged for it.
    3. This provides a way for users to go over their daily limit and get around many other policy rules. For example, if user A reached his max time on this tool for the day, he can ask user B to make a reservation and then transfer it to him.
    4. This doesn't check whether the receiving user is active, is qualified on the tool, has at least one active project, is even active on the same project that was chosen for the initial reservation (which is used to charge missed reservation) etc.

I am unfortunately not very optimistic about this being accepted by my managers, but I will bring it up at our next meeting.

**Comment by abuckles-uci** [comment]
Hi @rptmat57, thank you for this detailed response. We're very open to iterating and improving upon this PR to have it accepted.

File: NEMO/models.py [hunk 1]
```diff
@@ -2520,6 +2520,12 @@ def has_not_ended(self):
     def has_not_started(self):
         return False if self.start <= timezone.now() else True
 
+    def can_swap_user(self):
+        time_period = timedelta(hours=72)
+        if self.tool is not None and "TEM-" in self.tool.name:
```
Comment (arcturusannamalai) [inline NEMO/models.py:2525]: this seems to be the only equipment **hardcoded** string which is allowed for user swap; perhaps you can store this information in another table or update / migrate Models to add a boolean field to store this information?

**Comment by arcturusannamalai** [comment]
> Hello, Thank you for submitting this, the idea of letting users transfer their reservation to other users is an interesting one! I see this feature as pretty similar to reserving a tool on behalf of someone else (which currently only staff can do)
> 
> I have 2 main issues with the PR in its current state:
> 
> 1. As you said, there are a lot of hardcoded group names, times, rules etc. and I understand that those make sense for your facility but they probably would not make sense in any other.
> 2. There are a certain number of ways to break NEMO rules by using this feature, which you probably are ok with for your facility but this is unlikely to be the case in other places:
>    
>    1. None of the policy rules are checked and that could cause issues. Also the receiving user does not get a say in this transaction.
>    2. If the user doesn't know he is receiving the reservation, and misses it, he could potentially be charged for it.
>    3. This provides a way for users to go over their daily limit and get around many other policy rules. For example, if user A reached his max time on this tool for the day, he can ask user B to make a reservation and then transfer it to him.
>    4. This doesn't check whether the receiving user is active, is qualified on the tool, has at least one active project, is even active on the same project that was chosen for the initial reservation (which is used to charge missed reservation) etc.
> 
> I am unfortunately not very optimistic about this being accepted by my managers, but I will bring it up at our next meeting.

I think circumventing policy is a problem :/

**Pull request closed without merging**